### ISSUE TITLE
ECS structuring

### ISSUE THREAD

**Issue by fabiodr** [swift502/Sketchbook#30]
Great project! It expands a lot the vision of what kind of interactive experiences can be done on the web with Three.js.

Do you have thoughts about structuring with ECS pattern, using a lib like [ECSY](https://github.com/MozillaReality/ecsy) ?

Here is an ECSY example project also using typescript and physics: [https://github.com/macaco-maluco/thermal-runway](https://github.com/macaco-maluco/thermal-runway)

**Comment by swift502** [comment]
Interesting. Is that basically the Unity philosophy of "everything's an object within a scene with components attached to it"?

Would it maybe reduce the size of my character class and make it easier to navigate? Since it would just be an entity with components attached to it...?

On the other hand it seems like I'd be a slave to a strict project structure. Maybe that's a good thing but it's a bit scary.
Maybe after 0.4 is released, I'll look into this. But it seems like a lot of work. And there's some custom systems and philosophies in sketchbook (like character states), and I'm not entirely sure how they'll fit with the ESC pattern.

I'm putting this in the research phase, since I'm not that educated on how a ESC project looks like. I'll have to look into it more. 🙂

**Comment by fabiodr** [comment]
Yes, it’s like Unity in the sense that Components (MonoBehavior) can be shared between Entities (GameObjects) but there is a difference: behavior is defined in Systems, not in the Component itself.

It's much more similar to the new Unity DOTS, where data and operations/behavior are separated.

Components holds only data
Entity has components
Systems operate on entity's components

A System can specify a combination of Components it needs, then it filters and cache only Entities that have this combination to iterate them, implement the operation/behavior and transform the data in Components.

ECS favors reuse by composability over inheritance, it helps not to mix things that could be reused and to keep the operations focused and clean.

Some ECS libs in C++ (and also Unity DOTS in combination with Burst compiler) claims it improves even performance if used correctly, I don’t think it’s the case for JS but the overhead seems to be minimal looking at ECSY benchmarks.

In the long run, for medium to big complexity projects, I believe it helps with organization, reusability and maintainability.

As a reference for a future restructuration or experiment, Mozilla Hubs is large Three.js/Aframe application that is experimenting a migration to ECSY:
[https://github.com/mozilla/hubs/compare/feature/ecsy-integration](https://github.com/mozilla/hubs/compare/feature/ecsy-integration)
[https://github.com/mozilla/hubs/compare/ecsy-simple-entity-api](https://github.com/mozilla/hubs/compare/ecsy-simple-entity-api)
(I don't know if it goes further now with some sad recent Mozilla events)
